intake_esm/core.py

```diff
@@ -157,11 +162,19 @@ def serialize(self, name, directory=None):
             json_file_name = directory / json_file_name
 
         collection_data = self._col_data.copy()
-        collection_data['catalog_file'] = csv_file_name.as_posix()
         collection_data['id'] = name
 
-        print(f'Writing csv catalog to: {csv_file_name}')
-        self.df.to_csv(csv_file_name, compression='gzip', index=False)
+        if catalog_type == 'file':
+            collection_data['catalog_file'] = csv_file_name.as_posix()
+            print(f'Writing csv catalog to: {csv_file_name}')
+            self.df.to_csv(csv_file_name, compression='gzip', index=False)
+        else:
+            catalog_length = len(self.df)
+            print(f'Writing catalog with {catalog_length} entries into: {json_file_name}')
+            if catalog_length > 100:
+                print("  (call with catalog_type='file' to save catalog as a separate CSV file)")
+            collection_data['catalog_dict'] = self.df.to_dict(orient='records')
+
```

Comment:
Since `catalog_dict` and `catalog_file` are mutually exclusive, before finalizing the serialization process, we need to make sure that `catalog_data` doesn't contain both `catalog_dict` and `catalog_file`. Here's what I mean:

- If the original json file has `catalog_file`,  with this line `collection_data = self._col_data.copy()`, we end up getting `catalog_file` in `collection_data`. If the user then specifies `catalog_type='dict'`, we end up creating a new key `catalog_dict` and adding the catalog table to this. As a result, our json file now has both `catalog_dict` and `catalog_file`. 

One way to address this issue is to delete `catalog_dict`, `catalog_file` keys regardless of whether they are in the dictionary, 

```python
collection_data = self._col_data.copy()
for key in {'catalog_dict', 'catalog_file'}: 
    _ = catalog_data.pop(key, None)
```

After this, we can proceed to the next steps:

```python
collection_data['id'] = name
...
```


Example:                                     

```python
In [7]: catalog_data                                                                                               
Out[7]: {'catalog_dict': 'foo', 'other_key': 'bar'}

In [8]: for key in {'catalog_dict', 'catalog_file'}: 
   ...:     _ = catalog_data.pop(key, None) 
   ...:                                                                                                            

In [9]: catalog_data                                                                                               
Out[9]: {'other_key': 'bar'}
```

Reply:
These are good suggestions.   I will put these changes in shortly.